From f1f2bf4d8722ba53ac32c4cbf63614284a205fa5 Mon Sep 17 00:00:00 2001
From: kailua <kailua@project2by2.jp>
Date: Thu, 22 May 2025 18:44:05 +0900
Subject: [PATCH] eqe: Redo TEE-related blobs

* Pull strongbox feature from stock.
* Solve every dependencies
* Import weaver HAL
* Fixed TrickyStore
---
 ...> android.hardware.strongbox_keystore.xml} |  7 ++---
 device.mk                                     | 26 +++++++++++++++++--
 extract-files.py                              |  9 ++++++-
 proprietary-files.txt                         | 23 ++++++++++++++++
 4 files changed, 57 insertions(+), 8 deletions(-)
 rename configs/{android.hardware.strongbox_keystore_v4.xml => android.hardware.strongbox_keystore.xml} (85%)

diff --git a/configs/android.hardware.strongbox_keystore_v4.xml b/configs/android.hardware.strongbox_keystore.xml
similarity index 85%
rename from configs/android.hardware.strongbox_keystore_v4.xml
rename to configs/android.hardware.strongbox_keystore.xml
index 180e9ef2..c6ca188a 100644
--- a/device/motorola/eqe/configs/android.hardware.strongbox_keystore_v4.xml
+++ b/device/motorola/eqe/configs/android.hardware.strongbox_keystore.xml
@@ -1,20 +1,17 @@
 <?xml version="1.0" encoding="utf-8"?>
 <!-- Copyright (C) 2018 The Android Open Source Project
-
      Licensed under the Apache License, Version 2.0 (the "License");
      you may not use this file except in compliance with the License.
      You may obtain a copy of the License at
-
           http://www.apache.org/licenses/LICENSE-2.0
-
      Unless required by applicable law or agreed to in writing, software
      distributed under the License is distributed on an "AS IS" BASIS,
      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
      See the License for the specific language governing permissions and
      limitations under the License.
 -->
-
 <!-- Feature for devices with Keymaster in StrongBox. -->
 <permissions>
-    <feature name="android.hardware.strongbox_keystore" version="4" />
+  <feature name="android.hardware.strongbox_keystore" version="200"/>
+  <feature name="android.hardware.keystore.app_attest_key" />
 </permissions>
diff --git a/device.mk b/device.mk
index 062c1ee8..18c3426e 100644
--- a/device/motorola/eqe/device.mk
+++ b/device/motorola/eqe/device.mk
@@ -191,6 +191,10 @@ PRODUCT_COPY_FILES += \
 PRODUCT_PACKAGES += \
     android.hardware.biometrics.fingerprint@2.3-service.eqe
 
+# Gatekeeper
+PRODUCT_PACKAGES += \
+    android.hardware.gatekeeper@1.0.vendor
+
 # GPS
 PRODUCT_PACKAGES += \
     gnss@2.0-base.policy \
@@ -247,9 +251,23 @@ PRODUCT_COPY_FILES += \
 PRODUCT_COPY_FILES += \
     $(LOCAL_PATH)/keylayout/goodix_ts.kl:$(TARGET_COPY_OUT_VENDOR)/usr/keylayout/goodix_ts.kl
 
+# Keymaster
+PRODUCT_PACKAGES += \
+    android.hardware.keymaster@3.0.vendor \
+    android.hardware.keymaster@4.0.vendor \
+    android.hardware.keymaster@4.1.vendor \
+    lib_android_keymaster_keymint_utils.vendor \
+    libcppbor_external.vendor \
+    libkeymaster_messages.vendor \
+    libkeymaster_portable.vendor \
+    libpuresoftkeymasterdevice.vendor \
+    libsoft_attestation_cert.vendor
+
 # Keymint
 PRODUCT_PACKAGES += \
-    android.hardware.hardware_keystore.xml
+    android.hardware.hardware_keystore.xml \
+    android.hardware.security.keymint-V4-ndk.vendor \
+    libkeymint.vendor
 
 PRODUCT_COPY_FILES += \
     frameworks/native/data/etc/android.hardware.keystore.app_attest_key.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.hardware.keystore.app_attest_key.xml \
@@ -339,7 +357,7 @@ PRODUCT_COPY_FILES += \
     $(LOCAL_PATH)/configs/nfcee_access.xml:$(TARGET_COPY_OUT_SYSTEM)/etc/nfcee_access.xml
 
 PRODUCT_COPY_FILES += \
-    $(LOCAL_PATH)/configs/android.hardware.strongbox_keystore_v4.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.hardware.strongbox_keystore.xml
+    $(LOCAL_PATH)/configs/android.hardware.strongbox_keystore.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.hardware.strongbox_keystore.xml
 
 PRODUCT_COPY_FILES += \
     $(LOCAL_PATH)/configs/product_privapp-permissions-hotword.xml:$(TARGET_COPY_OUT_PRODUCT)/etc/permissions/privapp-permissions-hotword.xml
@@ -442,6 +460,10 @@ PRODUCT_COPY_FILES += \
 # VNDK
 PRODUCT_ENFORCE_ARTIFACT_PATH_REQUIREMENTS := strict
 
+# Weaver
+PRODUCT_PACKAGES += \
+    android.hardware.weaver@1.0.vendor
+
 # WiFi
 PRODUCT_PACKAGES += \
     android.hardware.wifi-service \
diff --git a/extract-files.py b/extract-files.py
index c0bd6768..220d00bd 100755
--- a/device/motorola/eqe/extract-files.py
+++ b/device/motorola/eqe/extract-files.py
@@ -76,8 +76,15 @@ def lib_fixup_vendor_suffix(lib: str, partition: str, *args, **kwargs):
         .replace_needed('android.hidl.base@1.0.so', 'libhidlbase.so'),
     'system_ext/lib64/libwfdservice.so': blob_fixup()
         .replace_needed('android.media.audio.common.types-V2-cpp.so', 'android.media.audio.common.types-V4-cpp.so'),
-    ('vendor/bin/hw/android.hardware.security.keymint-service-qti', 'vendor/lib64/libqtikeymint.so'): blob_fixup()
+    (
+        'vendor/bin/hw/android.hardware.security.keymint-service-qti',
+        'vendor/bin/hw/android.hardware.security.keymint-service-spu-qti',
+        'vendor/lib64/libqtikeymint.so',
+        'vendor/lib64/libspukeymint.so',
+    ): blob_fixup()
         .add_needed('android.hardware.security.rkp-V3-ndk.so'),
+    'vendor/lib64/libtpa.so': blob_fixup()
+        .replace_needed('android.hardware.security.keymint-V1-ndk.so', 'android.hardware.security.keymint-V4-ndk.so'),
     ('vendor/etc/media_codecs_crow_v0.xml', 'vendor/etc/media_codecs_crow_v1.xml', 'vendor/etc/media_codecs_crow_v2.xml'): blob_fixup()
         .regex_replace('.*media_codecs_(google_audio|google_c2|google_telephony|google_video|vendor_audio).*\n', ''),
     'vendor/etc/seccomp_policy/qwesd@2.0.policy': blob_fixup()
diff --git a/proprietary-files.txt b/proprietary-files.txt
index 4c121c3c..c0b3064e 100644
--- a/device/motorola/eqe/proprietary-files.txt
+++ b/device/motorola/eqe/proprietary-files.txt
@@ -791,7 +791,9 @@ vendor/lib64/vendor.goodix.hardware.biometrics.fingerprint@2.1.so
 
 # Gatekeeper
 vendor/bin/hw/android.hardware.gatekeeper@1.0-service-qti
+vendor/bin/hw/android.hardware.gatekeeper@1.0-service-spu-qti
 vendor/etc/init/android.hardware.gatekeeper@1.0-service-qti.rc
+vendor/etc/init/android.hardware.gatekeeper@1.0-service-spu-qti.rc
 vendor/lib64/hw/android.hardware.gatekeeper@1.0-impl-qti.so
 
 # GNSS
@@ -1032,17 +1034,32 @@ vendor/bin/vendor_modprobe.sh
 
 # Keymaster
 vendor/bin/hw/android.hardware.keymaster@4.0-service-qti
+vendor/bin/KmInstallKeybox
+vendor/bin/KmValidateKeybox
+vendor/bin/StoreKeybox
 vendor/lib64/libkeymasterdeviceutils.so
 vendor/lib64/libkeymasterprovision.so
 vendor/lib64/libkeymasterutils.so
+vendor/lib64/liboemcrypto.so
 vendor/lib64/libqtikeymaster4.so
+vendor/lib64/libtpa.so
 
 # Keymint
 vendor/bin/hw/android.hardware.security.keymint-service-qti
+vendor/bin/hw/android.hardware.security.keymint-service-spu-qti
+vendor/bin/hw/android.hardware.security.keymint-service.strongbox-thales
 vendor/etc/init/android.hardware.security.keymint-service-qti.rc
+vendor/etc/init/android.hardware.security.keymint-service.strongbox-thales.rc
 vendor/etc/vintf/manifest/android.hardware.security.keymint-service-qti.xml
+vendor/lib64/libjc_keymint-thales.so
+vendor/lib64/libjc_keymint_transport-thales.so
+vendor/lib64/libjc_keymint_transport_thales.so
 vendor/lib64/libqtikeymint.so
 vendor/lib64/libspcom.so
+vendor/lib64/libspukeymint.so
+vendor/lib64/libspukeymintdeviceutils.so
+vendor/lib64/libspukeymintprovision.so
+vendor/lib64/libspukeymintutils.so
 
 # Listen
 vendor/lib64/libcapiv2svacnnvendor.so
@@ -1745,6 +1762,12 @@ vendor/firmware/vpu30_4v_unsigned.mbn
 vendor/firmware/vpu33_4v.mbn
 vendor/firmware/vpu33_4v_unsigned.mbn
 
+# Weaver
+vendor/bin/hw/android.hardware.weaver-service-spu-qti
+vendor/bin/hw/android.hardware.weaver@1.0-service-thales
+vendor/etc/init/android.hardware.weaver@1.0-service-thales.rc
+vendor/lib64/ese_weaver_thales.so
+
 # Widevine DRM
 vendor/bin/hw/android.hardware.drm-service.widevine
 vendor/etc/init/android.hardware.drm-service.widevine.rc
